#**Scikit-learn**

###**Задание_1**

In [ ]:
import numpy as np
import pandas as pd

from sklearn.datasets import load_boston

In [ ]:
boston = load_boston()

feature_names = boston["feature_names"]

feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [ ]:
data = boston["data"]

data.shape

(506, 13)

In [ ]:
target = boston["target"]

target[:10]

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9])

In [ ]:
X = pd.DataFrame(data, columns = feature_names)

X.head()

CRIM    ZN  INDUS  CHAS    NOX  ...  RAD    TAX  PTRATIO       B  LSTAT
0  0.00632  18.0   2.31   0.0  0.538  ...  1.0  296.0     15.3  396.90   4.98
1  0.02731   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  396.90   9.14
2  0.02729   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  392.83   4.03
3  0.03237   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  394.63   2.94
4  0.06905   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  396.90   5.33

[5 rows x 13 columns]

In [ ]:
y = pd.DataFrame(target, columns = ["price"])

y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   price   506 non-null    float64
dtypes: float64(1)
memory usage: 4.1 KB


In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

In [ ]:
from sklearn.linear_model import LinearRegression

In [ ]:
lr = LinearRegression()

In [ ]:
lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
y_pred = lr.predict(X_test)

y_pred.shape

(152, 1)

In [ ]:
check_test = pd.DataFrame({
    "y_test": y_test["price"],
    "y_pred": y_pred.flatten(),
})

check_test.head(10)

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
76     20.0  23.146689
316    17.8  17.392124
140    14.0  14.078599
471    19.6  23.036927
500    16.8  20.599433

In [ ]:
from sklearn.metrics import r2_score

In [ ]:
r2_score(y_test, y_pred)

0.7112260057484974

###**Задание_2**

In [ ]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
model = RandomForestRegressor(n_estimators = 1000, max_depth = 12, random_state = 42)

In [ ]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=12, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=1000, n_jobs=None, oob_score=False,
                      random_state=42, verbose=0, warm_start=False)

In [ ]:
y_pred = model.predict(X_test)

y_pred.shape

(152,)

In [ ]:
check_test = pd.DataFrame({
    "y_test": y_test["price"],
    "y_pred": y_pred.flatten(),
})

check_test.head(10)

y_test     y_pred
173    23.6  22.846138
274    32.4  31.156114
491    13.6  16.297226
72     22.8  23.821036
452    16.1  17.212148
76     20.0  21.820092
316    17.8  19.866369
140    14.0  14.759938
471    19.6  21.235224
500    16.8  20.883103

In [ ]:
from sklearn.metrics import r2_score

In [ ]:
r2_score(y_test, y_pred)

0.8749965273218174

r2 "Случайный лес" = 0.7112260057484974 

r2 "Линейная регрессия" = 0.8749965273218174

"Случайный лес" работает лучше, чем "Линейная регрессия"

###**Задание_3**

In [ ]:
feature_importance = pd.DataFrame({'name':X.columns, 
                                   'feature_importance':model.feature_importances_}, 
                                  columns = ['feature_importance', 'name'])
feature_importance

feature_importance     name
0             0.032117     CRIM
1             0.001550       ZN
2             0.007094    INDUS
3             0.001149     CHAS
4             0.014368      NOX
5             0.402705       RM
6             0.014245      AGE
7             0.064033      DIS
8             0.004968      RAD
9             0.011692      TAX
10            0.018090  PTRATIO
11            0.012311        B
12            0.415679    LSTAT

In [ ]:
feature_importance.nlargest(2, 'feature_importance')

feature_importance   name
12            0.415679  LSTAT
5             0.402705     RM

Признаки LSTAT и RM наиболее важны